src/groups/semidirect_product_group.jl

```diff
@@ -88,296 +96,885 @@ See [HilgertNeeb:2012; Definition 9.2.22](@cite), first definition for more deta
-respectively, and ``σ`` is the group action specified by the [`AbstractGroupActionType`](#ref) within the [`LeftSemidirectProductLieGroup`](@ref)  ``L``.
+
+See also [`AbstractRightGroupActionType`](@ref) and [`ActionActsOnLeft`](@ref).
 """
 compose(
-    SDPG::LieGroup{𝔽, LeftSemidirectProductGroupOperation, <:ProductManifold}, ::Any, ::Any
-) where {𝔽}
+    SDPG::LieGroup{𝔽, <:SemidirectProductGroupOperation{O1, O2, A, AO}, <:ProductManifold}, ::Any, ::Any
+) where {𝔽, O1, O2, A <: AbstractRightGroupActionType, AO <: ActionActsOnLeft}
 
+# An implementation for 1,3 (no inverse for left) and 5,7 (inverse for right)
 function _compose!(
-        SDPG::LieGroup{𝔽, <:LeftSemidirectProductGroupOperation, <:ProductManifold}, k, g, h
-    ) where {𝔽}
-    PM = SDPG.manifold
-    G, H = map(LieGroup, PM.manifolds, SDPG.op.operations)
-    A = GroupAction(SDPG.op.action_type, G, H)
-    # We have to perform 3 steps applying the group action
-    # 1) x = σ_g[1](h[2]) (a point on H)
-    # 2) compose g[1] and h[1] (a point on G, neither input needed afterwards)
-    # 3) compose g[2] and x (a point on H, neither input needed afterwards)
-    # to avoid to overwrite elements in case k=g or k=h: allocate for result of (1)
-    # especially after (1) we still need g[2] (in case k=g)
-    x = copy(H, submanifold_component(SDPG, k, Val(2)))
-    # (1)
-    apply!(
-        A, x, submanifold_component(SDPG, g, Val(1)), submanifold_component(SDPG, h, Val(2))
-    )
-    # (2)
-    _compose!(
-        G,
-        submanifold_component(SDPG, k, Val(1)),
-        submanifold_component(SDPG, g, Val(1)),
-        submanifold_component(SDPG, h, Val(1)),
-    )
-    _compose!(
-        H, submanifold_component(SDPG, k, Val(2)), submanifold_component(SDPG, g, Val(2)), x
-    )
+        SDPG::LieGroup{𝔽, <:SemidirectProductGroupOperation{O1, O2, A, AO}, <:ProductManifold}, k, g, h
+    ) where {𝔽, O1, O2, A <: AbstractGroupActionType, AO <: ActionActsOnLeft}
+    PM, G, H, a, g_ind, h_ind = _semidirect_parts(SDPG)
+    if Base.mightalias(k, g) || Base.mightalias(h, g) # k/h may not overlap with g
+        kH = copy(H, submanifold_component(SDPG, k, Val(h_ind))) # to
+    else # if it does not alias, we can just use kG & kH
+        kH = submanifold_component(SDPG, k, Val(h_ind))
+    end
+    if Base.mightalias(k, h) || Base.mightalias(k, g) # h/g may not overlap with k
+        kG = copy(G, submanifold_component(SDPG, k, Val(g_ind))) # to
+    else # if it does not alias, we can just use kG & kH
+        kG = submanifold_component(SDPG, k, Val(g_ind))
+    end
+    # invert hG for right, copy for left
+    # this is inplace if both are not aliased and creates a copy kG otherwise to avoid overwriting hG
+    _semidirect_maybe_inv!(a, G, kG, submanifold_component(SDPG, h, Val(g_ind)))
+    # a) group action  (first to avoid side effects in g, set kH to σ_{kG}(gH), with the above this avoids aliasing
+    apply!(a, kH, kG, submanifold_component(SDPG, g, Val(h_ind)))
```

Comment:
Just double check this part please as I'm too tired to know for sure.
If `k === h`, `apply!` will modify `kH` and therefore `k[h_inx]`
So, I think ln 306 needs to also check for `Base.mightalias(k, h)`?